File: matlab/src/measure.m

function [params]=measure(params, fs)


iT = params.it;
time = params.dt * iT;
Efield = params.Efield;

% Loop over species
for s = 1:params.Ns
    % Get species-specific grid and distribution function
    grid = params.grids(s);
    f = fs(:,:,s);
    Vgrid = grid.V;

    % Calculate diagnostics
    Mass = sum(f, "all") * grid.dx * grid.dv;
    Momentum = sum(f .* Vgrid, "all") * grid.dx * grid.dv;
    Epot = 0.5 * sum(Efield.^2) * grid.dx;
    Ekin = 0.5 * sum(f .* (Vgrid.^2), "all") * grid.dx * grid.dv;
    Etot = Epot + Ekin;
    L2norm = sum(abs(f).^2, "all") * grid.dx * grid.dv;

    % Create a filename for the species
    species_name = params.species_name(s); % Species name
    filename = fullfile(params.data_dir, species_name+".csv");

    % Create a table row for the current measurement
    new_row = table(iT, time, Mass, Momentum, Epot, Ekin, Etot, L2norm, ...
        'VariableNames', {'it','time','Mass', 'Momentum', 'Epot', 'Ekin', 'Etot', 'L2norm'});

    % Check if the file already exists
    if exist(filename, 'file') && iT>1
        % Load the existing table
        existing_table = readtable(filename);
        % Append the new row
        updated_table = [existing_table; new_row];
    else
        % Create a new table with the current row
        updated_table = new_row;
        params.diagnostic_filename(s) = filename;
    end

    % Write the updated table to the file
    writetable(updated_table, filename);
end
end


function [Emode_abs] = fourier_modes(Efield, k_list, grid)
% this function implements (63-65) of https://arxiv.org/pdf/1009.3046.pdf
% for odd fourier numbers (k=0.5 etc)
ik = 1;
Ek = fft(Efield);
Emode_abs = abs(Ek(2:5))/grid.size(1);
%     for k = k_list
%         Esin_k = sum(Efield.*sin(2*pi*k*grid.x/grid.L(1)))*grid.dx;
%         Ecos_k = sum(Efield.*cos(2*pi*k*grid.x/grid.L(1)))*grid.dx;
%         Emode_abs(ik) = sqrt(abs(Esin_k)^2+abs(Ecos_k)^2)/grid.L(1);
%         ik = ik + 1;
%     end

end

File: src/measure.py
import jax.numpy as jnp
import pandas as pd
import os
import sys
from pathlib import Path

params_path = Path(__file__).resolve().parent.parent
print(params_path)
sys.path.append(str(params_path))

import params.PARAMS_ion_acoustic_waves as params

def measure(params, fs):
    iT = params.it
    time = params.dt * iT
    Efield = params.Efield
    
    for s in range(params.Ns):
        grid = params.grids[s]
        f = fs[:, :, s]             ## May not work
        Vgrid = grid['V']

        dx = grid['dx']
        dv = grid['dv']

        Mass = jnp.sum(f) * dx * dv
        Momentum = jnp.sum(f * Vgrid.T) * dx * dv
        Epot = 0.5 * jnp.sum(Efield ** 2) * dx
        Ekin = 0.5 * jnp.sum(f * (Vgrid.T ** 2)) * dx * dv
        Etot = Epot + Ekin
        L2norm = jnp.sum(jnp.abs(f) ** 2) * dx * dv

        species_name = params.species_name[s]
        filename = Path(params.data_dir) / f"{species_name}.csv"

        new_row = pd.DataFrame([{
            'it': iT,
            'time': time,
            'Mass': Mass,
            'Momentum': Momentum,
            'Epot': Epot,
            'Ekin': Ekin,
            'Etot': Etot,
            'L2norm': L2norm
        }])

        if filename.exists() and iT > 1:
            existing_table = pd.read_csv(filename)
            updated_table = pd.concat([existing_table, new_row], ignore_index=True)
        else:
            updated_table = new_row
            if not hasattr(params, 'diagnostic_filename'):
                params.diagnostic_filename = [None] * params.Ns
            params.diagnostic_filename[s] = str(filename)

        updated_table.to_csv(filename, index=False)

    return params


def fourier_modes(Efield, k_list, grid):
    # Implements basic version of modes based on FFT
    Ek = jnp.fft.fft(Efield)
    Emode_abs = jnp.abs(Ek[1:5]) / grid.size[0]
    
    return Emode_abs